### ISSUE TITLE
Make Link.Refresh more intelligent

### ISSUE THREAD

**Issue by mkromberg** [Dyalog/link#86]
Link.Refresh currently removes and re-creates a link, potentially doing a huge amount of unnecessary work. It should scan namespaces and folders for changes and only refresh items which have changed.

**Comment by nicolas-dyalog** [comment]
I don't see how scanning namespaces and folders and comparing them all can be significantly faster than just re-loading them all.

**Comment by mkromberg** [comment]
The performance hit will depending things like the cost of fragmenting the workspace, touching lots of pages in memory in a virtual machine, etc. As a general rule one should never do unnecessary work since the cost is fundamentally unknown. It is hard to believe that comparing can be SLOWER than re-loading.

More importantly, there are other reasons why we should not just re-load everything:

1) Re-fixing namespace and class definitions may execute code that the application only expects to have run once, at startup. This may be unavoidable if the definitions are changed, but should not be something that Link triggers if you do a refresh.
2) Re-fixing namespaces and classes will destroy their internal state - variables will be destroyed.
3) Re-fixing objects may leave existing refs to objects pointing to the old, now unnamed, definition. If editing is done after this, the situation may become chaotic.
4) Re-fixing things that are on the stack has been known to trigger bugs in the tracer and editor. Obviously, such bugs should be fixed, but there is no reason to provoke them unnecessarily.

If the same code is used for a "crawler" where a refresh is effectively done at regular intervals, the issues mentioned above will become intolerable.

It is a fundamental feature of Refresh that it has a minimal impact on the active workspace.

**Comment by abrudz** [comment]
Comparing can certainly be *much* faster than refixing, even in not too complex cases:
```
      ]runtime -c "(⎕SRC foo)≡⊃⎕NGET'/tmp/foo.apln'1" "⎕FIX'file:///tmp/foo.apln'"
                                                                                                
  (⎕SRC foo)≡⊃⎕NGET'/tmp/foo.apln'1 → 3.1E¯4 |      0%                                          
* ⎕FIX'file:///tmp/foo.apln'        → 8.2E¯2 | +26200% ⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕⎕ 
```
Where the foo is:
```
:Namespace foo
    :class goo1
        ∇ make
          :Implements constructor
          :Access public
        ∇
        ∇ make1(x)
          :Implements constructor
          :Access public
        ∇
        ∇ make2(x y)
          :Implements constructor
          :Access public
        ∇
        ∇ maken x
          :Implements constructor
          :Access public
        ∇
        :namespace inner
        :endnamespace
    :endclass
    :class goo2
        ∇ make1(x)
          :Implements constructor
          :Access public
        ∇
        ∇ make2(x y)
          :Implements constructor
          :Access public
        ∇
        ∇ maken x
          :Implements constructor
          :Access public
        ∇
        :namespace inner
        :endnamespace
    :endclass
    :class goo3
        ∇ make1(x)
          :Implements constructor
          :Access public
        ∇
        ∇ make2(x y)
          :Implements constructor
          :Access public
        ∇
        ∇ maken x
          :Implements constructor
          :Access public
        ∇
        :namespace inner
        :endnamespace
    :endclass
:EndNamespace
```

**Comment by nicolas-dyalog** [comment]
it's bug because refixing a namespace that doesn't need to be refixed may have unwanted side-effects.

**Comment by nicolas-dyalog** [comment]
Fixed in link v2.1.0-beta1
